### PR TITLE
 Add the ability to trigger a backup via a dedicated endpoint

### PR DESCRIPTION
**Pull request by thewhitewizard**

### CODE REVIEW

File: src/main/java/com/iexec/sms/admin/AdminController.java [hunk 1]
```diff
@@ -32,6 +32,20 @@
...
+     */
+    private static final String BACKUP_STORAGE_LOCATION = "/work/";
+
+    /**
+     * The name of the database backup file.
+     */
+    private static final String BACKUP_FILENAME = "backup.sql";
+
+    /**
+     * We want to perform one operation at a time. This  ReentrantLock is used to set up the lock mechanism.
```
Comment (mcornaton) [inline src/main/java/com/iexec/sms/admin/AdminController.java:47]: One unnecessary space there :)
```suggestion
     * We want to perform one operation at a time. This ReentrantLock is used to set up the lock mechanism.
```

File: src/main/java/com/iexec/sms/admin/AdminService.java [hunk 2]
```diff
@@ -15,15 +15,92 @@
...
+    /**
+     * Creates a backup of the H2 database and saves it to the specified location.
+     *
+     * @param storageLocation The location where the backup file will be saved, must be an existing directory.
+     * @param backupFileName  The name of the backup file.
+     * @return {@code true} if the backup was successful; {@code false} if any error occurs.
+     */
+    boolean createDatabaseBackupFile(String storageLocation, String backupFileName) {
+        // Check for missing or empty storageLocation parameter
+        if (storageLocation == null || storageLocation.isEmpty()) {
```
Comment (mcornaton) [inline src/main/java/com/iexec/sms/admin/AdminService.java:57]:
```suggestion
        if (StringUtils.isBlank(storageLocation)) {
```
?

File: src/main/java/com/iexec/sms/admin/AdminService.java [hunk 3]
```diff
@@ -15,15 +15,92 @@
...
+        }
+        // Ensure that storageLocation ends with a slash
+        if (!storageLocation.endsWith("/")) {
+            storageLocation += "/";
+        }
+
+        // Check if storageLocation is an existing directory, we don't want to create it.
+        final File directory = new File(storageLocation);
+        if (!directory.exists() || !directory.isDirectory()) {
+            log.error("storageLocation must be an existing directory.");
```
Comment (mcornaton) [inline src/main/java/com/iexec/sms/admin/AdminService.java:74]: Should we also log the `storageLocation` there?

File: src/main/java/com/iexec/sms/admin/AdminService.java [hunk 4]
```diff
@@ -15,15 +15,92 @@
...
+     */
+    boolean databaseDump(String fullBackupFileName) {
+
+        if (fullBackupFileName == null || fullBackupFileName.isEmpty()) {
+            log.error("fullBackupFileName must not be empty.");
+            return false;
+        }
+
+        try {
+            log.info("Starting the backup process {}", fullBackupFileName);
```
Comment (mcornaton) [inline src/main/java/com/iexec/sms/admin/AdminService.java:93]: Let's follow our log conventions :smile: 
```suggestion
            log.info("Starting the backup process [fullBackupFileName:{}]", fullBackupFileName);
```

File: src/main/java/com/iexec/sms/admin/AdminService.java [hunk 5]
```diff
@@ -15,15 +15,92 @@
...
+            final long start = System.currentTimeMillis();
+            Script.process(datasourceUrl, datasourceUsername, datasourcePassword, fullBackupFileName, "DROP", "");
+            final long stop = System.currentTimeMillis();
+            final long size = (new File(fullBackupFileName)).length();
+            log.info("New backup created [timestamp {}, duration {} ms, size {}]", dateFormat.format(new Date(start)), stop - start, size);
```
Comment (mcornaton) [inline src/main/java/com/iexec/sms/admin/AdminService.java:98]:
```suggestion
            log.info("New backup created [timestamp:{}, duration:{} ms, size:{}, fullBackupFileName:{}]", dateFormat.format(new Date(start)), stop - start, size, fullBackupFileName);
```

File: src/main/java/com/iexec/sms/admin/AdminService.java [hunk 6]
```diff
@@ -15,15 +15,92 @@
...
+        } catch (SQLException e) {
+            log.error("SQL error occurred during backup : " + e.getMessage());
```
Comment (mcornaton) [inline src/main/java/com/iexec/sms/admin/AdminService.java:100]: The following should print the full stacktrace:
```suggestion
            log.error("SQL error occurred during backup", e);
```

File: src/main/java/com/iexec/sms/admin/AdminController.java (hunk 1 above)
Comment (thewhitewizard) [reply src/main/java/com/iexec/sms/admin/AdminController.java:49]: done in local

File: src/main/java/com/iexec/sms/admin/AdminService.java (hunk 2 above)
Comment (thewhitewizard) [reply src/main/java/com/iexec/sms/admin/AdminService.java:57]: done in local

File: src/main/java/com/iexec/sms/admin/AdminService.java (hunk 3 above)
Comment (thewhitewizard) [reply src/main/java/com/iexec/sms/admin/AdminService.java:74]: log.error("storageLocation must be an existing directory [fullBackupFileName:{}]", storageLocation);

Comment (thewhitewizard) [reply src/main/java/com/iexec/sms/admin/AdminService.java:74]: log.error("storageLocation must be an existing directory [storageLocation:{}]", storageLocation);

File: src/main/java/com/iexec/sms/admin/AdminService.java (hunk 4 above)
Comment (thewhitewizard) [reply src/main/java/com/iexec/sms/admin/AdminService.java:93]: done in local

File: src/main/java/com/iexec/sms/admin/AdminService.java (hunk 5 above)
Comment (thewhitewizard) [reply src/main/java/com/iexec/sms/admin/AdminService.java:98]: done in local

File: src/main/java/com/iexec/sms/admin/AdminService.java (hunk 6 above)
Comment (thewhitewizard) [reply src/main/java/com/iexec/sms/admin/AdminService.java:100]: done in local

File: src/main/java/com/iexec/sms/admin/AdminService.java [hunk 7]
```diff
@@ -71,7 +71,7 @@ boolean createDatabaseBackupFile(String storageLocation, String backupFileName)
         // Check if storageLocation is an existing directory, we don't want to create it.
         final File directory = new File(storageLocation);
         if (!directory.exists() || !directory.isDirectory()) {
```
[the remark this answers is not in the document]
Comment (thewhitewizard) [reply src/main/java/com/iexec/sms/admin/AdminService.java:73]: done in local

**Review by mcornaton — APPROVED** [review]

File: src/main/java/com/iexec/sms/admin/AdminService.java [hunk 8]
```diff
@@ -35,8 +45,62 @@ public AdminService(@Value("${spring.datasource.url}") String datasourceUrl,
...
+        return databaseDump(storageLocation + backupFileName);
+    }
+
+    /**
+     * @param fullBackupFileName complete fileName (location and filename)
+     * @return {@code true} if the backup was successful; {@code false} if any error occurs.
+     */
+    boolean databaseDump(String fullBackupFileName) {
+
+        if (fullBackupFileName == null || fullBackupFileName.isEmpty()) {
```
Comment (jbern0rd) [inline src/main/java/com/iexec/sms/admin/AdminService.java:87]: For consistency with `createDatabaseBackupFile` implementation
```suggestion
        if (StringUtils.isBlank(fullBackupFileName)) {
```

**Review by jbern0rd — APPROVED** [review]
Thanks for this implementation

File: src/main/java/com/iexec/sms/admin/AdminService.java [hunk 9]
```diff
@@ -35,8 +45,62 @@ public AdminService(@Value("${spring.datasource.url}") String datasourceUrl,
...
+            return false;
+        }
+        // Check for missing or empty backupFileName parameter
+        if (StringUtils.isBlank(backupFileName)) {
+            log.error("backupFileName must not be empty.");
+            return false;
+        }
+        // Ensure that storageLocation ends with a slash
+        if (!storageLocation.endsWith("/")) {
+            storageLocation += "/";
```
Comment (jbern0rd) [inline src/main/java/com/iexec/sms/admin/AdminService.java:68]: There is a `File.separator` constant value which is nice to know for such usages

File: src/main/java/com/iexec/sms/admin/AdminService.java [hunk 10]
```diff
@@ -35,8 +45,62 @@ public AdminService(@Value("${spring.datasource.url}") String datasourceUrl,
...
+            log.error("fullBackupFileName must not be empty.");
+            return false;
+        }
+
+        try {
+            log.info("Starting the backup process [fullBackupFileName:{}]", fullBackupFileName);
+            final long start = System.currentTimeMillis();
+            Script.process(datasourceUrl, datasourceUsername, datasourcePassword, fullBackupFileName, "DROP", "");
+            final long stop = System.currentTimeMillis();
+            final long size = (new File(fullBackupFileName)).length();
```
Comment (jbern0rd) [inline src/main/java/com/iexec/sms/admin/AdminService.java:97]: I think it should be allowed to write this
```suggestion
            final long size = new File(fullBackupFileName).length();
```

Comment (thewhitewizard) [reply src/main/java/com/iexec/sms/admin/AdminService.java:97]: done in local

File: src/main/java/com/iexec/sms/admin/AdminService.java (hunk 9 above)
Comment (thewhitewizard) [reply src/main/java/com/iexec/sms/admin/AdminService.java:68]: done in local, it's clearly better

File: src/main/java/com/iexec/sms/admin/AdminService.java (hunk 8 above)
Comment (thewhitewizard) [reply src/main/java/com/iexec/sms/admin/AdminService.java:87]: yes, I had made the change before I saw your comment

**Pull request merged**